mlflow/models/evaluation/__init__.py

```diff
@@ -0,0 +1,399 @@
+            md5_gen.update(EvaluationDataset._array_like_obj_to_bytes(data))
+        else:
+            md5_gen.update(
+                EvaluationDataset._array_like_obj_to_bytes(
+                    data[: EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH]
+                )
+            )
+            md5_gen.update(
+                EvaluationDataset._array_like_obj_to_bytes(
+                    data[-EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH:]
+                )
+            )
+
+    @property
+    def name(self):
+        return self._user_specified_name if self._user_specified_name is not None else self.hash
+
+    @property
+    def hash(self):
+        """
+        Compute a hash from the specified dataset by selecting the first 5 records, last 5 records,
+        dataset size and feeding them through a cheap, low-collision hash function
+        """
+        if self._hash is None:
+            md5_gen = hashlib.md5()
+            if isinstance(self.data, np.ndarray):
+                EvaluationDataset._gen_md5_for_arraylike_obj(md5_gen, self.data)
+                EvaluationDataset._gen_md5_for_arraylike_obj(md5_gen, self.labels)
+            elif isinstance(self.data, pd.DataFrame):
+                EvaluationDataset._gen_md5_for_arraylike_obj(md5_gen, self.data)
+                md5_gen.update(self.labels.encode("UTF-8"))
+            self._hash = md5_gen.hexdigest()
+        return self._hash
+
+    @property
+    def _metadata(self):
+        metadata = {
+            "name": self.name,
+            "hash": self.hash,
+        }
+        if self.path is not None:
+            metadata["path"] = self.path
+        return metadata
+
+
+class ModelEvaluator:
+    def can_evaluate(self, model_type, evaluator_config=None, **kwargs) -> bool:
+        """
+        :param model_type: A string describing the model type (e.g., "regressor",
+                           "classifier", …).
+        :param evaluator_config: A dictionary of additional configurations for
+                                 the evaluator.
+        :param **kwargs: For forwards compatibility, a placeholder for additional
+                         arguments that may be added to the evaluation interface
+                         in the future.
+        :return: True if the evaluator can evaluate the specified model on the
+                 specified dataset. False otherwise.
+        """
+        raise NotImplementedError()
+
+    def compute_metrics_and_compute_and_log_artifacts(
+        self, model, model_type, dataset, evaluator_config, run_id
+    ):
+        """
+        return an tuple of:
+         - an instance of EvaluationMetrics
+         - a dict of artifact_name -> instance_of_EvaluationArtifact
+        and log artifacts into run specified by run_id
+        """
+        raise NotImplementedError()
+
+    def evaluate(
+            self,
+            model: PyFuncModel,
+            model_type,
+            dataset,
+            run_id=None,
```

Comment:
From https://github.com/mlflow/mlflow/pull/5069/files#r755590594, `run_id` should never be `None`.

Reply:
+1

Reply:
I refactor the logic, add a context manager with verification:
```
class StartRunOrReuseActiveRun:

    def __init__(self, run_id):
        self.user_specified_run_id = run_id
        self.managed_run = None

    def __enter__(self):
        if mlflow.active_run() is not None:
            active_run_id = mlflow.active_run().info.run_id
            if self.user_specified_run_id is not None and \
                    self.user_specified_run_id != active_run_id:
                raise ValueError("An active run exists, you cannot specify another run_id when "
                                 "evaluating.")
            return active_run_id
        else:
            if self.user_specified_run_id is None:
                raise ValueError("Active run does not exist, you need specify a run_id when "
                                 "evaluating.")
            self.managed_run = \
                mlflow.start_run(run_id=self.user_specified_run_id).__enter__()
            return self.user_specified_run_id

    def __exit__(self, exc_type, exc_val, exc_tb):
        if self.managed_run is not None:
            return self.managed_run.__exit__(exc_type, exc_val, exc_tb)
```